src/core/selinux-setup.c

```diff
@@ -104,3 +107,38 @@ int mac_selinux_setup(bool *loaded_policy) {
 
         return 0;
 }
+
+int mac_selinux_setup_namespaced_policy(const char *policy) {
+
+#if HAVE_SELINUX
+        int r;
+        _cleanup_close_ int fd = -EBADF;
+
+        assert(policy);
+
+        // TODO use selinux_unshare() from libselinux
+        fd = open("/sys/fs/selinux/unshare", O_WRONLY | O_CLOEXEC);
+        if (fd < 0)
+                return errno;
+
+        r = RET_NERRNO(write(fd, "1\n", sizeof("1\n")));
+        if (r < 0)
+                return r;
+
+        r = RET_NERRNO(umount("/sys/fs/selinux"));
+        if (r < 0)
+                return r;
+
+        r = RET_NERRNO(mount("selinuxfs", "/sys/fs/selinux", "selinuxfs", MS_NODEV|MS_NOEXEC|MS_NOSUID, NULL));
+        if (r < 0)
+                return r;
+
+        int enforce = 0;
+        r = selinux_init_load_policy(&enforce);
```

Comment:
so, this is run after chrooting into the target dir, i.e. we are highly privileged typically, but we read the policy from a potentially untrusted root fs. Is this properly safe?

Reply:
No, the policy should be loaded from the PID 1 namespace.